Userland/Libraries/LibTest/Randomized/Generator.h

```diff
@@ -0,0 +1,321 @@
+//
+// Gen::one_of(20,5,10) --> value 20, RandomRun [0]
+//                      --> value 5,  RandomRun [1]
+//                      --> value 10, RandomRun [2]
+//
+// Shrinks towards the earlier arguments (above, towards 20).
+template<typename... Ts>
+requires(sizeof...(Ts) > 0)
+CommonType<Ts...> one_of(Ts... choices)
+{
+    Vector<CommonType<Ts...>> choices_vec { choices... };
+
+    constexpr size_t count = sizeof...(choices);
+    size_t i = unsigned_int(count - 1);
+    return choices_vec[i];
+}
+
+template<typename T>
+struct Choice {
+    i32 weight;
+    T value;
+};
+
+// Randomly (uniformly) selects a value out of the given weighted arguments.
+//
+// Gen::frequency(
+//   Gen::Choice{5,'x'},
+//   Gen::Choice{1,'o'}
+// )
+//     --> value 'x' (5 out of 6 times), RandomRun [0]
+//     --> value 'o' (1 out of 6 times), RandomRun [1]
+//
+// Shrinks towards the earlier arguments (above, towards 'x').
+template<typename... Ts>
+requires(sizeof...(Ts) > 0)
+CommonType<Ts...> frequency(Choice<Ts>... choices)
+{
+    Vector<Choice<CommonType<Ts...>>> choices_vec { choices... };
+
+    u32 sum = 0;
+    for (auto const& choice : choices_vec) {
+        VERIFY(choice.weight > 0);
+        sum += static_cast<u32>(choice.weight);
+    }
+
+    u32 target = unsigned_int(sum);
+    size_t i = 0;
+    for (auto const& choice : choices_vec) {
+        u32 weight = static_cast<u32>(choice.weight);
+        if (weight >= target) {
+            return choice.value;
+        }
+        target -= weight;
+        ++i;
+    }
+    return choices_vec[i - 1].value;
+}
+
+// An unsigned integer generator in the full u32 range.
+//
+// 8/17 (47%) of the time it will bias towards 8bit numbers,
+// 4/17 (23%) towards 4bit numbers,
+// 2/17 (12%) towards 16bit numbers,
+// 1/17 (6%) towards 32bit numbers,
+// 2/17 (12%) towards edge cases like 0 and NumericLimits::max() of various unsigned int types.
+//
+// Gen::unsigned_int() -> value 3,     RandomRun [0,3]
+//                     -> value 8,     RandomRun [1,8]
+//                     -> value 100,   RandomRun [2,100]
+//                     -> value 5,     RandomRun [3,5]
+//                     -> value 255,   RandomRun [4,1]
+//                     -> value 65535, RandomRun [4,2]
+//                     etc.
+//
+// Shrinks towards 0.
+inline u32 unsigned_int()
+{
+    u32 bits = frequency(
+        // weight, bits
+        Choice { 4, 4 },
+        Choice { 8, 8 },
+        Choice { 2, 16 },
+        Choice { 1, 32 },
+        Choice { 2, 0 });
+
+    // The special cases go last as they can be the most extreme (large) values.
+
+    if (bits == 0) {
+        // Special cases, eg. max integers for u8, u16, u32.
+        return one_of(
+            0U,
+            NumericLimits<u8>::max(),
+            NumericLimits<u16>::max(),
+            NumericLimits<u32>::max());
+    }
+
+    u32 max = (bits == 32)
+        ? NumericLimits<u32>::max()
+        : ((u64)1 << bits) - 1;
```

Comment:
Why are we special-casing `u32` here? `u64` should be big enough to handle the "common" case, even for a 32-bit value.

Reply:
I didn't catch that. Which u32 are you talking about?

Taking a guess at what you mean - the first u32 in this snippet. I need `u32 max` because `u32 unsigned_int(u32)` takes u32. I could try to make `unsigned_int` return u64s, perhaps by `RandomRun` being built on `u64` or by simply taking two `u32`s of randomness instead of one.

Is that what you're talking about?

Reply:
Limiting everything to `u32` is fine. The question is why this is making a special case for `bits == 32` instead of just being `u32 max = ((u64)1 << bits) - 1` without this distinction? There is no issue with shifting an `u64` by 32 bits, so that shouldn't be needed at all.

Reply:
Ah, gotcha! I'll simplify that, thanks for the explanation.